stl/inc/shared_mutex

```diff
@@ -227,46 +279,85 @@ public:
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared()) {} // construct with mutex and try to lock shared
 
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, adopt_lock_t) noexcept // strengthened
+    _NODISCARD_CTOR_LOCK
+    _Requires_lock_held_(_Mtx)
+    _Post_satisfies_(_Owns == true)
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, adopt_lock_t) noexcept // strengthened
         : _Pmtx(_STD addressof(_Mtx)), _Owns(true) {} // construct with mutex and adopt ownership
 
     template <class _Rep, class _Period>
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, const chrono::duration<_Rep, _Period>& _Rel_time)
+    _NODISCARD_CTOR_LOCK
+    _Requires_lock_not_held_(_Mtx)
+    _When_(this->_Owns == true, _Acquires_shared_lock_(_Mtx))
+    _When_(this->_Owns == true, _Acquires_nonreentrant_lock_(_Mtx))
+    _When_(this->_Owns == true, _Post_same_lock_(_Mtx, *this->_Pmtx))
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, const chrono::duration<_Rep, _Period>& _Rel_time)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared_for(_Rel_time)) {
         // construct with mutex and try to lock for relative time
     }
 
     template <class _Clock, class _Duration>
-    _NODISCARD_CTOR_LOCK shared_lock(mutex_type& _Mtx, const chrono::time_point<_Clock, _Duration>& _Abs_time)
+    _NODISCARD_CTOR_LOCK
+    _Requires_lock_not_held_(_Mtx)
+    _When_(this->_Owns == true, _Acquires_shared_lock_(_Mtx))
+    _When_(this->_Owns == true, _Acquires_nonreentrant_lock_(_Mtx))
+    _When_(this->_Owns == true, _Post_same_lock_(_Mtx, *this->_Pmtx))
+    _Post_satisfies_(_Pmtx == &_Mtx)
+    shared_lock(mutex_type& _Mtx, const chrono::time_point<_Clock, _Duration>& _Abs_time)
         : _Pmtx(_STD addressof(_Mtx)), _Owns(_Mtx.try_lock_shared_until(_Abs_time)) {
         // construct with mutex and try to lock until absolute time
 #if _HAS_CXX20
         static_assert(chrono::is_clock_v<_Clock>, "Clock type required");
 #endif // _HAS_CXX20
     }
 
+    _When_(this->_Owns == true, _Releases_shared_lock_(*this->_Pmtx))
+    _When_(this->_Owns == true, _Releases_nonreentrant_lock_(*this->_Pmtx))
     ~shared_lock() noexcept {
         if (_Owns) {
             _Pmtx->unlock_shared();
         }
     }
 
-    _NODISCARD_CTOR_LOCK shared_lock(shared_lock&& _Other) noexcept : _Pmtx(_Other._Pmtx), _Owns(_Other._Owns) {
+    _NODISCARD_CTOR_LOCK
+    _Moves_lock_(*this->_Pmtx, *_Other._Pmtx)
+    _Post_satisfies_(_Owns == _Other._Owns)
+    _Post_satisfies_(_Other._Pmtx == nullptr)
+    _Post_satisfies_(_Other._Owns == false)
+    shared_lock(shared_lock&& _Other) noexcept : _Pmtx(_Other._Pmtx), _Owns(_Other._Owns) {
         _Other._Pmtx = nullptr;
         _Other._Owns = false;
     }
 
+    _Releases_shared_lock_(*this->_Pmtx)
+    _Releases_nonreentrant_lock_(*this->_Pmtx)
     shared_lock& operator=(shared_lock&& _Right) noexcept {

```

Comment:
Open issue: I don't believe that this is correct. This is a move assignment operator, but is being annotated as releasing only. From https://learn.microsoft.com/en-us/cpp/code-quality/annotating-locking-behavior?view=msvc-170 it appears that `_Replaces_lock_` should be involved, especially as the move constructor immediately above uses `_Moves_lock_`.

Reply:
Agree. Will update.

Reply:
Fixed.

Reply:
I changed to `_Moves_lock_` here but seems like compiler doesn’t understand it when used with `std::move`.